text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+Overall, FPC can be considered an addition to Fabric wherein all chaincode computation relies only on the correctness of data provided by an authenticated Client or computed inside of and signed by a validated Trusted Execution Environment. It assumes that the Ordering Service is a trusted system element. The Endorsing Peer outside of the Trusted Execution Environment, however, is considered untrusted.
+
+Writing chaincode for FPC should come natural to developers familiar with Fabric as the programming model (e.g., chaincode lifecycle, chaincode invocations and state) is the same as for normal Fabric chaincode.  The main differences are a (for now at least) different programming language (C++) and a Shim API which, on the one hand, is somewhat simplified but, on the other hand, also contains FPC-specific extensions such as allowing for private as well as public ledger state as well as a secure (yet “replicable”) random-number generator.
+See the [FPC Shim](#fpc-shim) section for more information.
+
+In general, a developer can write chaincode largely without having to be aware that the code executes within a Trusted Execution Environment (TEE), often referred to as an Enclave. However, to understand the architecture and security underpinning it we introduce in the following several new concepts and terms. Except where noted otherwise, all elements of the architecture described below reside on a Fabric Peer.
+
+
+## Terminology:
+
+* Attestation: An important cryptographic feature of Trusted Execution Environments by which the hardware can reliably state exactly what software it is running. This statement is signed by the hardware so that anyone reading it can verify that the statement came from an actual valid and up-to-date TEE. The Attestation can cover both the software portion of the TEE itself as well as any program to be run in it. This attestation takes the form of a "quote" containing (among other things) a measurement of the code and data in the TEE and its software version number, which is signed by the TEE and can be verified by anyone using well known Public Key Infrastructure technology.
+
+* Trusted Execution Environment (TEE): The isolated secure environment in which programs run in encrypted memory, unreadable even by privileged users or system processes. FPC chaincodes run in TEEs.
+
```

Comment:
Perhaps add a section that describes the capabilities / primitives of the TEE, such as:

The TEE can:
- Produce a signature that every other TEE can verify
- Store with integrity up to `n` values in persistent storage

Reply:
Good point that we should be more explicit (e.g., as part of the TEE entry in the terminology section) in what properties we need in a TEE.  The key properties besides the already mentioned "encrypted memory, unreadable even by privileged users or system processes" is the ability to provide (publicly verifiable!) attestation and seal state (i.e., encrypt data with a key which is known only to a particular enclave).  These latter two properties is what enables, e.g., the two bullets you mention above.